src/plugins/discover/public/application/main/components/sidebar/lib/get_data_view_field_list.ts

```diff
@@ -7,18 +7,32 @@
  */
 
 import { difference } from 'lodash';
-import { DataView, DataViewField } from '@kbn/data-views-plugin/public';
+import type { DataView, DataViewField } from '@kbn/data-views-plugin/public';
+import { fieldWildcardFilter } from '@kbn/kibana-utils-plugin/public';
 import { isNestedFieldParent } from '../../../utils/nested_fields';
 
-export function getDataViewFieldList(dataView?: DataView, fieldCounts?: Record<string, number>) {
-  if (!dataView || !fieldCounts) return [];
+export function getDataViewFieldList(
+  dataView: DataView | undefined,
+  fieldCounts: Record<string, number> | undefined | null,
+  isPlainRecord: boolean
+) {
+  const currentFieldCounts = fieldCounts || {};
+  const sourceFiltersValues = dataView?.getSourceFiltering?.()?.excludes;
+  let dataViewFields: DataViewField[] = dataView?.fields.getAll() || [];
 
-  const fieldNamesInDocs = Object.keys(fieldCounts);
-  const fieldNamesInDataView = dataView.fields.getAll().map((fld) => fld.name);
+  if (sourceFiltersValues) {
+    const filter = fieldWildcardFilter(sourceFiltersValues, dataView.metaFields);
+    dataViewFields = dataViewFields.filter((field) => {
+      return filter(field.name) || currentFieldCounts[field.name]; // don't filter out a field which was present in hits (ex. for text-based queries)
+    });
+  }
```

Comment:
I think for text based query languages we can skip the field filtering, FYI @stratoula